### PR TITLE
refactor(lit-form): deprecate AbstractModel methods that require BinderNode

### LINKED ISSUE #2060: [lit-form] make model API not depend on form binder
In few cases, the current models' API implementation requires direct access to the respective binder node:
- `AbstractModel::valueOf()` - alternative: `binderNode.value` 
- `AbstractModel::toString()` - alternative: `String(binderNode.value)`
- `ArrayModel::[Symbol.iterator]()` - no alternative yet

As part of introducing [Object-based models](https://github.com/vaadin/hilla/issues/1903), and extracting models into a separate package for wider use, it would be great to replace these APIs with some alternatives that do not require invoking binder nodes from models.

Let's deprecate these APIs and provide alternative API for iterating arrays in the binder node.

Documentation needs to be updated accordingly.

### PR DESCRIPTION
**Pull request by Lodin**
Resolves #2060.

### CODE REVIEW

**Review by platosha — APPROVED** [review]

**Review by platosha — CHANGES_REQUESTED** [review]

File: packages/ts/lit-form/src/Models.ts [hunk 1]
```diff
@@ -213,40 +230,54 @@ export class ArrayModel<MItem extends AbstractModel = AbstractModel> extends Abs
...
+   * @deprecated Use the {@link getIterator} function instead.
+   */
+  *[Symbol.iterator](): IterableIterator<BinderNode<MItem>> {
+    for (const item of this[_items]()) {
       yield getBinderNode(item);
     }
   }
 }
+
+export function getIterator<M extends ArrayModel>(model: M): Generator<ArrayItemModel<M>, void, void> {
```
[the remark this answers is not in the document]
Comment (Lodin) [reply packages/ts/lit-form/src/Models.ts:281]: Hmm, I usually expect to have an array by the function named this way. Could also be confusing... Maybe, `getItemsIterator` / `getItemsIter`? Or `iterateItems`?

File: packages/ts/lit-form/src/Models.ts [hunk 2]
```diff
@@ -213,39 +230,53 @@ export class ArrayModel<MItem extends AbstractModel = AbstractModel> extends Abs
...
+  }
+
+  /**
+   * Iterates over the current model and yields a binder node for every item
+   * model.
+   *
+   * @deprecated Use the {@link m.items} function instead. E.g.:
+   * ```ts
+   * const itemValues = Array.from(m.items(model), (itemModel) => m.value(itemModel));
+   * ```
```
Comment (platosha) [inline packages/ts/lit-form/src/Models.ts:276]: `m.value` is not available yet :-)

Maybe it's fine to use some use-case specific examples for Lit and React:

```suggestion
   * @deprecated Use the {@link m.items} function instead. For example, in React:
   * ```tsx
   * const {model, field} = useForm(GroupModel);
   * return Array.from(m.items(model.people), (personModel) => 
   *   <TextField label="Full name" {...field(personModel.fullName)} />
   * );
   * ```
   * In Lit:
   * ```ts
   * return html`${repeat(m.items(this.binder.model.people), personModel => html`
   *   <vaadin-text-field label="Full name" ${field(personModel.fullName)}></vaadin-text-field>
   *  `)}
   * `;
```

File: packages/ts/lit-form/src/index.ts [hunk 3]
```diff
@@ -7,6 +9,20 @@ export * from './Validation.js';
...
+   *
+   * @param model - The array model to iterate over.
+   * @returns An iterator over item models.
+   */
+  items<M extends ArrayModel>(model: M): Generator<ArrayItemModel<M>, void, void> {
+    return model[_items]() as Generator<ArrayItemModel<M>, void, void>;
+  },
+};
+
+export default m;
```
Comment (platosha) [inline packages/ts/lit-form/src/index.ts:24]: Let's not use default export from the package index here — it's the Form library, not the models library. Does not feel semantically correct to export the model utils as the default asset here.

**Review by platosha — CHANGES_REQUESTED** [review]

File: packages/ts/lit-form/src/index.ts (hunk 3 above)
Comment (Lodin) [reply packages/ts/lit-form/src/index.ts:24]: Ah, right, didn't think about that. Thanks

**Review by cromoteca — APPROVED** [review]

**Pull request merged**